server/plat/rtcache.cpp

```diff
@@ -19,74 +19,75 @@
-     }
-     else
-     {
-         printf("Invalid App Name \n");
-     }
-
-}
+#define printf(...) printf(__VA_ARGS__); \
+                    fflush(stdout);
 
 rtError rtAppStatusCache::UpdateAppStatusCache(rtValue app_status)
 {
+     std::unique_lock<std::mutex> lock(CacheMutex);
      printf("RTCACHE : %s\n",__FUNCTION__);
 
       rtError err;
       rtObjectRef temp = app_status.toObject();
 
-      const char *App_name = strdup(temp.get<rtString>("applicationName").cString());
-      printf("App Name = %s\nApp ID = %s\nApp State = %s\nError = %s\n",App_name,temp.get<rtString>("applicationId").cString(),temp.get<rtString>("state").cString(),temp.get<rtString>("error").cString());
-
-      std::string id = getAppCacheId(App_name);
+      std::string cache_name_id = std::string(temp.get<rtString>("applicationName").cString());
+      printf("App Name = %s\nApp ID = %s\nApp State = %s\nError = %s\n",cache_name_id.c_str(),temp.get<rtString>("applicationId").cString(),temp.get<rtString>("state").cString(),temp.get<rtString>("error").cString());
 
-      if(doIdExist(id)) {
+      if(doIdExist(cache_name_id)) {
           printf("erasing old data\n");
-          err = ObjectCache->erase(id);
+          err = ObjectCache->erase(cache_name_id);
       }
 
-      err = ObjectCache->insert(id,temp);
+      err = ObjectCache->insert(cache_name_id, temp);
+      CacheCondVar.notify_all(); 
       return err;
 }
 
-const char * rtAppStatusCache::SearchAppStatusInCache(const char *app_name)
+bool rtAppStatusCache::WaitForAppState(const char * app_name, const char * desired_state, unsigned int timeout_ms) {
+     bool result = false;
+     bool wait_for_stopped = (strcmp(desired_state, "stopped") == 0);
+     std::unique_lock<std::mutex> lock(CacheMutex);
+     unsigned int time_left = timeout_ms;
+     unsigned long time_elapsed = 0;
+     std::chrono::steady_clock::time_point start_time = std::chrono::steady_clock::now();
+     printf("RTCACHE : %s Enter waiting application: %s state: %s\n", __FUNCTION__, app_name, desired_state);
+     while (time_elapsed < timeout_ms) {
+          const char * state = SearchAppStatusInCacheLocked(app_name);
+          if (strcmp(state, "NOT_FOUND") == 0) {
+               printf("RTCACHE : %s application: %s Not found\n", __FUNCTION__, app_name);
+               if (wait_for_stopped) {
+                    result = true;
+                    printf("RTCACHE : %s waiting for stopped - leave\n", __FUNCTION__);
+                    break;
+               }
+          }
+          if (strcmp(state, desired_state) == 0) {
+               printf("RTCACHE : %s desired state: %s, application: %s - leave\n", __FUNCTION__, desired_state, app_name);
+               result = true;
+               break;
+          }
+          printf("RTCACHE : %s Waiting ... %u ms\n", __FUNCTION__, time_left);
+          CacheCondVar.wait_for(lock, std::chrono::milliseconds(time_left));
```

Comment:
Is it vulnerable to spurious awakenings?

Reply:
You may noticed or not ... but ... it is surrounded  by while loop with checking the time .... It could wakeup, but will recalculate time left to wait, check the state and break if desired state  is achieved